core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java

```diff
@@ -226,16 +228,68 @@ public void getStatus(GetStatusPRequest request,
   @Override
   public void listStatus(ListStatusPRequest request,
       StreamObserver<ListStatusPResponse> responseObserver) {
-    String path = request.getPath();
-    ListStatusPOptions options = request.getOptions();
-    RpcUtils.call(LOG, (RpcUtils.RpcCallableThrowsIOException<ListStatusPResponse>) () -> {
-      List<FileInfo> result = new ArrayList<>();
-      for (alluxio.wire.FileInfo fileInfo : mFileSystemMaster.listStatus(new AlluxioURI(path),
-          ListStatusContext.create(options.toBuilder()))) {
-        result.add(GrpcUtils.toProto(fileInfo));
+
+    // Receive {@link alluxio.wire.FileInfo} items from master.
+    List<alluxio.wire.FileInfo> fileInfoList;
+    try {
+      fileInfoList = RpcUtils.callAndReturn(LOG,
+          () -> mFileSystemMaster.listStatus(new AlluxioURI(request.getPath()),
+              ListStatusContext.create(request.getOptions().toBuilder())),
+          "ListStatus", false, "request: %s", request);
+    } catch (StatusException se) {
+      responseObserver.onError(se);
+      return;
+    }
+
+    // Stream results back in chunks.
+    final int resultPacketLength =
+        ServerConfiguration.getInt(PropertyKey.MASTER_FILE_SYSTEM_LISTSTATUS_RESULT_PACKET_LENGTH);
+
+    // Abort condition for packet streaming.
+    SettableFuture<Void> abortCondition = SettableFuture.create();
+    // Current position in the master fileInfo list.
+    int currentIdx = 0;
+    do {
+      // Abort the packet if cancellation was signaled in previous package.
+      if (abortCondition.isCancelled()) {
+        break;
       }
-      return ListStatusPResponse.newBuilder().addAllFileInfos(result).build();
-    }, "ListStatus", "path=%s, options=%s", responseObserver, path, options);
+
+      // How many remaining items to stream.
+      int remainingItemCount = fileInfoList.size() - currentIdx;
+      // Start index in the master fileInfo list for the next packet.
+      int packetStartIdx = currentIdx;
+      // How many results to stream in this round.
+      int packetItemCount = Math.min(resultPacketLength, remainingItemCount);
+      // Whether this is the last packet.
+      boolean isLastPacket = (remainingItemCount == packetItemCount);
+
+      String packetDescription =
```

Comment:
Can we put description arguments at the end of `streamingRPCAndLog` call to defer message formatting to only when it is logged?